### PR TITLE
conformace/run_e2e.sh: pass shellcheck and autodetect ginkgo nodes

### PR DESCRIPTION
**Pull request by neolit123**
**What this PR does / why we need it**:
- for ginkgo parallel mode pass -p to autodetect ginkgo nodes.
- disable parallel mode by default (false).
- use provider "skeleton" instead of "local".
- make run_e2e.sh pass shellcheck.

**Which issue(s) this PR fixes**:

xref https://github.com/kubernetes/kubernetes/issues/74853

**Special notes for your reviewer**:
NONE

**Does this PR introduce a user-facing change?**:

```release-note
NONE
```

/kind bug
/priority important-longterm
/assign @dims @BenTheElder

### CODE REVIEW

File: cluster/images/conformance/run_e2e.sh
Comment (spiffxp) [inline cluster/images/conformance/run_e2e.sh]: WDYT about this for parity (y or numbers seems weird to me)
```suggestion
    'n'|'N')           GINKGO_NODES=1 ;;
    'y'|'Y')           GINKGO_NODES=25 ;;
```

File: cluster/images/conformance/run_e2e.sh
Comment (spiffxp) [inline cluster/images/conformance/run_e2e.sh]: I suspect there are other taints you might want (IIRC hybrid windows/linux clusters are supposed to be tainting their windows nodes out) but I won't block for trying this out

Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: copy pasted from @BenTheElder 's code.

File: cluster/images/conformance/run_e2e.sh
Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: @dims WDYT?
this reflects in the YAML.

**Comment by spiffxp** [comment]
/approve
/hold
if you want to address the comments (and I'll let someone else /lgtm)

File: cluster/images/conformance/run_e2e.sh
Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: i spoke to sig-windows about hybrid and there is a tracking issue somewhere, but not sure if this will land this cycle. last time i've checked there was a blocker in the e2e framework on this front.

Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: @spiffxp `70431`

**Comment by neolit123** [comment]
> Some packages in hack/.shellcheck_failures are passing shellcheck. Please remove them.
>  ./cluster/images/conformance/run_e2e.sh

will do, after we decide on: https://github.com/kubernetes/kubernetes/pull/74854#discussion_r261839715

File: cluster/images/conformance/run_e2e.sh
Comment (BenTheElder) [reply cluster/images/conformance/run_e2e.sh]: My code only needed to handle kind clusters specifically, so we're depending on kubeadm's behavior + our own. Not sure if this is sufficient for all clusters.

Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: i guess that's true. someone might decide to allow scheduling workloads on their CP nodes.
will try to propose something more generic.

**Comment by neolit123** [comment]
@dims @spiffxp @BenTheElder 

update:
- remove file from the list of files to ignore shell check
- added proposal of detecting workload-able nodes (waiting feedback on this)

File: cluster/images/conformance/run_e2e.sh
Comment (neolit123) [reply cluster/images/conformance/run_e2e.sh]: side question here: isn't the default of 25 for parallel a bit too much?

**Comment by neolit123** [comment]
> pull-kubernetes-conformance-image-test Pending — Job triggered.

```
I0303 17:02:59.574]  ---> 2a8f0665106f
I0303 17:02:59.574] Step 10/14 : ENV E2E_PROVIDER="local"
I0303 17:02:59.858]  ---> Running in 7ed189ade0e0
I0303 17:03:00.210] Removing intermediate container 7ed189ade0e0
I0303 17:03:00.210]  ---> 51550c0ca395
I0303 17:03:00.210] Step 11/14 : ENV E2E_PARALLEL="1"
```
huh, this is not running on this PR.

/test pull-kubernetes-conformance-image-test

File: cluster/images/conformance/run_e2e.sh
Comment (timothysc) [inline cluster/images/conformance/run_e2e.sh]: So I like everything here, but let's take out this part and put it in the framework.

**Review by timothysc — CHANGES_REQUESTED** [review]
minor change request, but everything else +1 

/cc @stevesloka

**Comment by neolit123** [comment]
updated to not include `--num-nodes` detection.
it's to be part of the e2e FW:
https://github.com/kubernetes/kubernetes/pull/74905

updated PR title too.

/test pull-kubernetes-conformance-image-test

**Comment by dims** [comment]
cc @johnSchnake @brahmaroutu

**Review by timothysc — APPROVED** [review]
/lgtm 
/approve 
/hold 

for final comments ^

File: cluster/images/conformance/run_e2e.sh
Comment (waynr) [inline cluster/images/conformance/run_e2e.sh]: Removing this breaks `sonobuoy`'s `--e2e-parallel` flag which previously let users like me exercise our best judgement in choosing test execution parallelism. Now the only way to increase parallelism for me is to spin up a VM with larger vcpu count.

File: cluster/images/conformance/run_e2e.sh
Comment (waynr) [inline cluster/images/conformance/run_e2e.sh]: This is somewhat naive. It means any conformance tests marked with `[Serial]` won't be run, which reduces the value of the conformance suite. That being said, I have noticed that the number of such tests has significantly reduced (if i remember correctly, in kubernetes 1.11.x it was around 7).

Comment (BenTheElder) [reply cluster/images/conformance/run_e2e.sh]: You cannot safely run `[Serial]` tagged tests in parallel, that's what the tag means. ginkgo nodes must be zero to run these.

File: cluster/images/conformance/run_e2e.sh
Comment (waynr) [inline cluster/images/conformance/run_e2e.sh]: I suspect this newfangled way of constructing the `ginkgo` command line has consequences for how special characters in `E2E_SKIP` AND `E2E_FOCUS` need to be escaped since the same values for these variables that worked on older kubernetes versions (using the old heptio conformance image) cause my conformance tests to run 0 tests total. This combined with the changes in this PR to the way `E2E_PARALLEL` is handled was sufficient to fool me into thinking that my tests were passing since `sonobuoy e2e <tarball>` only shows the number of failed tests, nothing about the total number of tests run.

File: cluster/images/conformance/run_e2e.sh
Comment (waynr) [reply cluster/images/conformance/run_e2e.sh]: I think "naive" was probably a bad choice of words, I was actually thinking of "misleading". It's misleading because it allows users to think they have run all conformance tests, when in reality there may be some `[Serial]` tests that haven't been run. Maybe I'm just overreacting though. Anyway, previously for CI purposes I would run `sonobuoy` twice -- once for all the tests that could be run in parallel, and again for all the tests that had to be run serially.

Comment (BenTheElder) [reply cluster/images/conformance/run_e2e.sh]: For submitting conformance results the standard is one run in non-serial mode / a single ginkgo node with a specific config, but doing something like this is definitely very useful for CI results. :+1:

**Comment by neolit123** [comment]
@waynr
> Removing this breaks sonobuoy's --e2e-parallel flag which previously let users like me exercise our best judgement in choosing test execution parallelism. Now the only way to increase parallelism for me is to spin up a VM with larger vcpu count.

are you requesting a revert or would you prefer to submit a fix on top of this PR?
cc @johnSchnake

**Comment by waynr** [comment]
@neolit123 I think I'll submit a PR with a fix for this if that's okay. I'll try to do that by the end of the day.

**Comment by neolit123** [comment]
sure go ahead.
i'm just wondering why this was reported so late after the PR merged.

**Comment by waynr** [comment]
Cause I only just started digging into the different sonobuoy behaviors and ran across this... I thought it might be useful for the folks who originally submitted and reviewed this PR to learn about the downstream impact.

**Comment by neolit123** [comment]
understood, letting the users pick the value makes perfect sense and is the finer grain control, but perhaps we should also allow them to say "please auto-detect this for me".

**Pull request merged**
